src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,112 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_FACTOR = 1.5;
+
+    private int size;
+    private T[] elementData;
+
+    public ArrayList() {
+        this(DEFAULT_CAPACITY);
+    }
+
+    public ArrayList(int initialCapacity) {
+        this.elementData = (T[]) new Object[initialCapacity];
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity(size + 1);
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexForAdd(index);
+        ensureCapacity(size + 1);
+        if (index != size) {
+            System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        }
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int requiredCapacity = size + list.size();
+        ensureCapacity(requiredCapacity);
+        for (int i = 0; i < list.size(); i++) {
+            elementData[size++] = list.get(i);
+        }
```

Comment:
I think that we can reuse the previous method add(int index). It will be better if we are talking about the DRY principle
```suggestion
        for (int i = 0; i < list.size(); i++) {
            add(list.get(i));
        }
```

Reply:
Thanks, agree